File: backends/fastapi/app/models/capsule.py
from __future__ import annotations

import uuid
from datetime import datetime, timezone

from sqlalchemy import Boolean, ForeignKey, Integer, String, Text, Uuid
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base
from app.db.types import UTCDateTime


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


class Capsule(Base):
    __tablename__ = "capsules"

    id: Mapped[uuid.UUID] = mapped_column(Uuid, primary_key=True, default=uuid.uuid4)
    owner_id: Mapped[uuid.UUID] = mapped_column(
        Uuid, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    code: Mapped[str] = mapped_column(String(8), nullable=False, unique=True, index=True)
    title: Mapped[str] = mapped_column(String(60), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    open_at: Mapped[datetime] = mapped_column(UTCDateTime, nullable=False, index=True)
    in_plaza: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    favorite_count: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    created_at: Mapped[datetime] = mapped_column(
        UTCDateTime, nullable=False, default=_utcnow, index=True
    )
    updated_at: Mapped[datetime] = mapped_column(
        UTCDateTime, nullable=False, default=_utcnow, onupdate=_utcnow
    )

File: fullstacks/rails/app/models/capsule.rb
class Capsule < ApplicationRecord
  self.table_name = "capsules"
  self.primary_key = "id"

  cross_db_uuid :id, :owner_id
  cross_db_timestamp :open_at, :created_at, :updated_at
  # in_plaza 在 SQLite 是 INTEGER 0/1、在 PG 是 boolean——统一转 Ruby 布尔（0 在 Ruby 为真值，必须转）。
  attribute :in_plaza, :boolean

  before_create { self.id = SecureRandom.uuid if id.blank? }
end
